src/errors.rs

```diff
@@ -0,0 +1,508 @@
+    Diff(#[from] DiffError),
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    Fetch(#[from] FetchError),
+}
+
+//////////////////////////////////////////////////////////
+// DiffError
+//////////////////////////////////////////////////////////
+
+#[derive(Debug, Error, Diagnostic)]
+#[non_exhaustive]
+pub enum DiffError {
+    #[error("Failed to diff package")]
+    CommandError(
+        #[from]
+        #[diagnostic(cause)]
+        CommandError,
+    ),
+    #[error("failed to acquire permission to diff")]
+    DiffDenied(
+        #[from]
+        #[diagnostic(cause)]
+        tokio::sync::AcquireError,
+    ),
+}
+
+//////////////////////////////////////////////////////////
+// UserInfoError
+//////////////////////////////////////////////////////////
+
+#[derive(Debug, Error, Diagnostic)]
+#[non_exhaustive]
+pub enum UserInfoError {
+    #[error("Failed to get user.name")]
+    UserCommandFailed(#[diagnostic(cause)] CommandError),
+    #[error("Failed to get user.email")]
+    EmailCommandFailed(#[diagnostic(cause)] CommandError),
+}
+
+//////////////////////////////////////////////////////////
+// FetchError
+//////////////////////////////////////////////////////////
+
+#[derive(Debug, Error, Diagnostic)]
+#[non_exhaustive]
+pub enum FetchError {
+    #[error("Invalid URL for package: {url}")]
+    InvalidUrl {
+        url: String,
+        #[diagnostic(cause)]
+        error: url::ParseError,
+    },
+    #[error("Running as --frozen but needed to fetch {package}:{version}")]
+    Frozen {
+        package: PackageName,
+        version: Version,
+    },
+    #[error("Failed to unpack .crate at {}", src.display())]
+    Unpack {
+        src: PathBuf,
+        #[diagnostic(cause)]
+        error: UnpackError,
+    },
+    #[error("failed to open cached .crate at {}", target.display())]
+    OpenCached {
+        target: std::path::PathBuf,
+        #[diagnostic(cause)]
+        error: std::io::Error,
+    },
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    Download(#[from] DownloadError),
+}
+
+#[derive(Debug, Error, Diagnostic)]
+#[non_exhaustive]
+pub enum UnpackError {
+    #[error("Failed to iterate archive")]
+    ArchiveIterate(#[diagnostic(cause)] std::io::Error),
+    #[error("Failed to read archive entry")]
+    ArchiveEntry(#[diagnostic(cause)] std::io::Error),
+    #[error("Invalid archive, {} wasn't under {}", entry_path.display(), prefix.to_string_lossy())]
+    InvalidPaths {
+        entry_path: PathBuf,
+        prefix: OsString,
+    },
+    #[error("Failed to unpack archive entry {}", entry_path.display())]
+    Unpack {
+        entry_path: PathBuf,
+        #[diagnostic(cause)]
+        error: std::io::Error,
+    },
+    #[error("Failed to finalize unpack to {}", target.display())]
+    LockCreate {
+        target: std::path::PathBuf,
+        #[diagnostic(cause)]
+        error: std::io::Error,
+    },
+    #[error(transparent)]
+    IoError(#[from] std::io::Error),
```

Comment:
I'm not sure that we need to have this much granularity for errors when unpacking crates, we can probably get away with coalescing them somewhat as it's not like the user can do anything about a package downloaded from crates.io being broken.